README.md
# Hierarcy
 2d haskell strategy game


CHANGELOG.md
# Changelog for `newFolder`

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to the
[Haskell Package Versioning Policy](https://pvp.haskell.org/).

## Unreleased

## 0.1.0.0 - YYYY-MM-DD


app/Main.hs
module Main (main) where
import Graphics.Gloss
import Graphics.Gloss.Interface.IO.Interact
import Graphics.Gloss.Interface.Environment
import GraphicsUtils 
import Data.Maybe
import Data.List
import Subject
import Data.ByteString (ByteString, pack)


paths = ["D:\\Hierarchy\\images\\"++show i ++".bmp" | i <-  [1..9]]
imgsTrans imgs = [translate dx 0 img | (img,dx) <- zip imgs [0,300..9*300]] :: [Picture]



updateScreenState :: Event -> ScreenState -> ScreenState
updateScreenState (EventKey (MouseButton WheelUp) Down _ _) screen_state = screen_state{subjectInCenter = subjects screen_state !! newIndex}
        where newIndex =( 1 + fromMaybe 0 (findIndex (==subjectInCenter screen_state) $ subjects screen_state) ) `mod` len
              len = (length $ subjects screen_state) :: Int
updateScreenState (EventKey (MouseButton WheelDown) Down _ _) screen_state = screen_state{subjectInCenter = subjects screen_state !! newIndex}
        where newIndex =( -1 + fromMaybe 0 (findIndex (==subjectInCenter screen_state) $ subjects screen_state) ) `mod` len
              len = (length $ subjects screen_state) :: Int
--updateScreenState (EventKey (MouseButton LeftButton) Down _ (x,y)) screen_state = screen_state {subjectInCenter = subj}
--        where subj = subject $ fromMaybe (SubjectButton (subjectInCenter screen_state)) $ clickedObject x y screen_state
updateScreenState (EventMotion (x,y)) screen_state = screen_state {subjectInCenter = subj}
        where subj = subject $ fromMaybe (SubjectButton (subjectInCenter screen_state)) $ clickedObject x y screen_state
updateScreenState (EventKey (Char 't') Down _ _) screen_state = screen_state {subjectInCenter = Subject "" Blank, focusTarget = SubjectInCenter}
updateScreenState (EventKey (Char 'e') Down _ _) screen_state = screen_state {focusTarget = SubjectsAround}
updateScreenState (EventKey (Char 'r') Down _ _) screen_state = screen_state {focusTarget = Balanced}
updateScreenState _ screen_state = screen_state

                                



--initSubjects = take 7 $ cycle $ map (\s -> Subject [s] Blank) ['a','b'..'z'] :: [Subject]
--colors = map (dark)  [red,green,rose, yellow, chartreuse, rose,  chartreuse, aquamarine, azure, violet]
bitmapData = pack $ take 40000 (cycle [200,10,10,55])
main :: IO ()
main = do 
        imgs <- mapM loadBMP paths
        (width, height) <- getScreenSize
        putStrLn $ "width: " ++ (show width) ++ " height: " ++ (show height)
        play FullScreen (greyN 0.3) 60  (ScreenState width height (mainCharacter (imgs !! 0) ) (initSubjects $ tail imgs) (mainCharacter (imgs !! 0)) SubjectInCenter) drawScreen updateScreenState (\_ -> id)
        --display FullScreen (dark blue) ( coloredDrawing width height  ) 
        --display FullScreen (white) (bitmapOfByteString 100 100 (BitmapFormat TopToBottom PxRGBA) bitmapData True)
        --display FullScreen (white) (pictures $ imgsTrans imgs)
        where
                initSubjects imgs = take 7 $  cycle $ zipWith (\s img -> Subject s img) combinations (cycle imgs) :: [Subject]
                combinations = [ [c,b,a] | a <- ['a'..'z'], b <- ['a'..'z'], c <- ['a'..'z']]
                mainCharacter img = Subject "Main Character" img


app/WorkingMain1.hs
module WorkingMain1 (main) where
import Graphics.Gloss
import Graphics.Gloss.Data.Vector
import Graphics.Gloss.Interface.Environment
background :: Color
background = white


-- TODO : MOVE TO GraphicsUtils.hs ------------------
drawing :: Int -> Int -> Int -> Picture
drawing width height n  = color (greyN 0.4) $ scale scaleX scaleY  $ separateScreen radiusV n
       where 
              area = fromIntegral $ width * height :: Float
              radiusV =  ( area / (2*pi*(fromIntegral n+1)) ) ** (1/2) :: Float
              scaleX = (fromIntegral width / fromIntegral height) ** (1/2)
              scaleY = 1/scaleX
              

separateScreen radius n = rotate (90) $ pics
       where 
              pics = pictures (sCircle : lines)
              sCircle = circleSolid radius 
              lines = map (\i -> line [(0,0),mulSV 1000 $ ithCosSin i]) [0..n-1]
              radians = 2*(pi::Double) /  fromIntegral n
              ithPoint i = mulSV radius $ ithCosSin i
              ithCosSin i = (realToFrac $ cos $ radians * fromIntegral i,realToFrac $ sin $ radians * fromIntegral i) 


-----------------------------------------------------



main :: IO ()
main =  do 
              (width, height) <- getScreenSize
              putStrLn $ "width: " ++ (show width) ++ " height: " ++ (show height)
              display FullScreen cyan (drawing width height 7)

              

app/WorkingMain2.hs
module WorkingMain2 (main) where
import Graphics.Gloss
import Graphics.Gloss.Data.Vector
import Graphics.Gloss.Interface.Environment
background :: Color
background = white


-- TODO : MOVE TO GraphicsUtils.hs ------------------
drawing :: Int -> Int -> Int -> Picture
drawing width height n  = color (greyN 0.4) $ scale scaleX scaleY $ separateScreen radiusV n
       where 
              area = fromIntegral $ width * height :: Float
              radiusV =  ( area / (2*pi*(fromIntegral n+1)) ) ** (1/2) :: Float
              scaleX = (fromIntegral width / fromIntegral height) ** (1/2)
              scaleY = 1/scaleX
              


separateScreen radius n = rotate (90) $ pics
       where 
              pics =  pictures $ circles ++ lines
              circles =  centerCircle : coloredSectors
              thArc =       let    r = 400 
                                   thickness = r - l * cos (realToFrac radians/2)
                            in thickArc 0 degrees r (2*thickness)
              coloredSectors  =   zipWith (color) colors sectors
              sectors = zipWith (<>) littleCircles thickArcs
              ithThickArc i = rotate (degrees * fromIntegral (negate i) ) thArc -- (4 - i) stands for correct segment indexes
              thickArcs = map ithThickArc [0..n-1]
              degrees = 360/fromIntegral n
              centerCircle = circleSolid radius 
              l = radius / (1 - sin (realToFrac radians/2 )) :: Float
              littleCircle = circleSolid (l * sin (realToFrac radians/2))
              littleCircles = zipWith (\(x,y) -> translate x y) (map ithCircleCenter [0..n-1]) (replicate n littleCircle)
              lines = map (\i -> line [(0,0),mulSV 1000 $ ithCosSin i]) [0..n-1]
              radians = 2*(pi::Double) /  fromIntegral n
              ithCosSin i = (realToFrac $ cos $ radians * fromIntegral i,realToFrac $ sin $ radians * fromIntegral i) 
              ithCircleCenter i = mulSV l $ ithCosSinForCircles i
              ithCosSinForCircles i = (realToFrac $ cos $ radians *(1/2+ fromIntegral i) ,realToFrac $ sin $ radians *(1/2+ fromIntegral i) ) 


-----------------------------------------------------

colors = [red,green,blue,yellow, magenta, rose, orange, chartreuse, aquamarine, azure, violet]

main :: IO ()
main =  do 
              (width, height) <- getScreenSize
              putStrLn $ "width: " ++ (show width) ++ " height: " ++ (show height)
              display FullScreen cyan (drawing width height (length colors))

              

app/WorkingMain3.hs
module WorkingMain3 (main) where
import Graphics.Gloss
import Graphics.Gloss.Data.Vector
import Graphics.Gloss.Interface.Environment
import System.IO.Unsafe
import Data.List
import Data.Word
import Data.ByteString (ByteString, pack)


-- TODO : MOVE TO GraphicsUtils.hs ------------------
drawing :: Int -> Int -> Int -> Picture
drawing width height n  = color rose $ scale scaleX scaleY $ separateScreen $ ActorsPositionsOnScreen radiusV n --radiusV n
       where 
              area = fromIntegral $ width * height :: Float
              radiusV = ( area / (2*pi*(fromIntegral n+1)) ) ** (1/2)  :: Float
              scaleX = (fromIntegral width / fromIntegral height) ** (1/2)
              scaleY = 1/scaleX
              
data ActorsPositionsOnScreen = ActorsPositionsOnScreen {radius :: Float, parts_quantity :: Int} deriving (Eq, Show)
distanceToLittleCircle :: ActorsPositionsOnScreen -> Float
distanceToLittleCircle actsPosOnScreen =  radius' / (1 - sin (realToFrac radians/2 ))
       where
              radius' = radius actsPosOnScreen
              radians = 2*(pi::Double) /  fromIntegral (parts_quantity actsPosOnScreen)
arcThickness :: ActorsPositionsOnScreen -> Float
arcThickness actsPosOnScreen  =2* ( arcRadius - l * cos (realToFrac radians/2))
       where 
              l = distanceToLittleCircle actsPosOnScreen
              radians = 2*(pi::Double) /  fromIntegral (parts_quantity actsPosOnScreen)
arcRadius ::  Float
arcRadius = 10000



separateScreen actsPosOnScreen = rotate (90) $ pics
       where 
              radius' = radius actsPosOnScreen
              n = parts_quantity actsPosOnScreen
              pics =  pictures $ circles -- ++ lines    ------> not needed yet
              --lines = map (\i -> line [(0,0),mulSV 1000 $ ithCosSin i]) [0..n-1]
              circles =    coloredSectors ++ [centerCircle] 
              coloredSectors  = drawSectors actsPosOnScreen
              degrees = 360/fromIntegral n
              centerCircle = circleSolid (radius' )
              radians = 2*(pi::Double) /  fromIntegral n
              ithCosSin i = (realToFrac $ cos $ radians * fromIntegral i,realToFrac $ sin $ radians * fromIntegral i) 


drawSectors actsPosOnScreen = coloredSectors
       where  
              radius' = radius actsPosOnScreen
              n = parts_quantity actsPosOnScreen
              coloredSectors  =   zipWith (color) colors sectors
              sectors = zipWith (<>) littleCircles thickArcs
              radians = 2*(pi::Double) /  fromIntegral n
              littleCircle = circleSolid (l * sin (realToFrac radians/2))
              littleCircles = zipWith (\(x,y) -> translate x y) (map ithCircleCenter [0..n-1]) (replicate n littleCircle)
              l =  distanceToLittleCircle actsPosOnScreen--radius' / (1 - sin (realToFrac radians/2 )) :: Float
              degrees = 360/fromIntegral n
              thickArcs = map ithThickArc [0..n-1]
              ithThickArc i = rotate (degrees * fromIntegral (negate i) ) thArc
              ithCircleCenter i = mulSV l $ ithCosSinForCircles i
              ithCosSinForCircles i = (realToFrac $ cos $ radians *(1/2+ fromIntegral i) ,realToFrac $ sin $ radians *(1/2+ fromIntegral i) ) 
              thArc = thickArc 0 degrees arcRadius (arcThickness actsPosOnScreen)
-----------------------------------------------------

colors = map (dark)  [red,green,rose, yellow, chartreuse, rose,  chartreuse, aquamarine, azure, violet]
bitmapData = pack $ take 40000 (cycle [200,10,10,55])
main :: IO ()
main = do 
              imgs <- mapM loadBMP paths
              (width, height) <- getScreenSize
              putStrLn $ "width: " ++ (show width) ++ " height: " ++ (show height)
              display FullScreen (dark blue) (drawing width height (length colors))
              --display FullScreen (white) (bitmapOfByteString 100 100 (BitmapFormat TopToBottom PxRGBA) bitmapData True)
              --display FullScreen (white) (pictures $ imgsTrans imgs)
       where 
              paths = ["images\\"++show i ++".bmp" | i <-  [1..9]]
              imgsTrans imgs = [translate (i*300) 0 img | (img,i) <- zip imgs [1..9]] :: [Picture]


src/GraphicsUtils.hs
module GraphicsUtils (drawScreen
                     ,ScreenState(..)
                     ,FocusTarget(..)
                     ,segments_quantity
                     ,distanceToLittleCircle
                     ,littleCircleRadius
                     ,segmentInDegrees
                     ,segmentInRadians
                     ,centerCircleRadius
                     ,polar2decart
                     ,decart2polar
                     ,ClickableObject(..)
                     ,clickedObject) where
import Graphics.Gloss
import Graphics.Gloss.Data.Vector
import Graphics.Gloss.Interface.Environment
import Data.List
import Data.Word
import Data.Maybe
import Data.List
import Subject
import Data.ByteString (ByteString, pack)

data ClickableObject = SubjectButton { subject :: Subject}



clickedObject :: Float -> Float -> ScreenState -> Maybe ClickableObject
clickedObject x_ y_ screen_state= fmap (\s -> SubjectButton s) subj
       where
              (dx,dy) = focusPoint screen_state
              (x,y) = ( 1 / scaleX screen_state * x_+dx , 1 / scaleY screen_state * y_+dy )
              --subj = (subjects screen_state) !! ((index +1) `mod` segments_quantity screen_state)
              subj = if x**2+y**2 < (centerCircleRadius screen_state) ** 2 then  Just (Subject "" Blank)
                     else fmap (\i -> (subjects screen_state) !! ((i +1) `mod` segments_quantity screen_state)) index
              index = (findIndex (\(xc,yc) -> (x-xc)**2 + (y-yc)**2 <= (littleCircleRadius screen_state)**2) 
                     [ polar2decart (distanceToLittleCircle screen_state, -pi/2 +(1/2 + fromIntegral i ) * ( segmentInRadians screen_state) )  
                     | i <- [1..segments_quantity screen_state]]  )

border_angle  = 1/2 :: Float
arcRadius = 10000 :: Float

data FocusTarget = SubjectInCenter | SubjectsAround | Balanced deriving (Eq, Show)

data ScreenState = ScreenState {width :: Int, height :: Int, main_character :: Subject , subjects :: [Subject], subjectInCenter :: Subject, focusTarget :: FocusTarget } deriving (Eq, Show)

scaleX :: ScreenState -> Float
scaleX screen_state =  (fromIntegral (width screen_state) / fromIntegral (height screen_state)) ** (1/2)
scaleY screen_state =  (fromIntegral (height screen_state) / fromIntegral (width screen_state) ) ** (1/2)

colors = map (dark) $  cycle [greyN 0.3]--,red,green,rose, yellow,chartreuse, rose,  chartreuse, aquamarine, azure, violet]

drawScreen :: ScreenState-> Picture
drawScreen  screen_state@(ScreenState width height  _ _ subj_in_center _) = scale (scaleX screen_state) (scaleY screen_state) $ translate (-1*x) (-1*y)$ separateScreen screen_state -- 
       where 
              (x,y) = focusPoint screen_state
              n = segments_quantity screen_state
              area = fromIntegral $ width * height :: Float
              radiusV = ( area / (2*pi*(fromIntegral n+1)) ) ** (1/2)  :: Float
              
focusPoint screen_state = if subj_in_center_idx < 0 then (0,0) else  (x,y) --(x * scaleY screen_state,y * scaleX screen_state)
       where 
              (x,y) = polar2decart (distanceToLittleCircle screen_state,-pi/2 + (segmentInRadians screen_state) *(1/2+ fromIntegral subj_in_center_idx))
              subj_in_center_idx = fromMaybe (-1) (findIndex (==subj_in_center) $ subjects screen_state)
              subj_in_center = subjectInCenter screen_state

separateScreen screen_state = rotate (90) $ pictures $ sectors
       where 
              sectors =  coloredSectors ++ [drawCenterCircle screen_state] 
              coloredSectors  = drawSectors screen_state


              
              --ithCosSin i = polar2decart (1,i*radians) --(realToFrac $ cos $ radians * fromIntegral i,realToFrac $ sin $ radians * fromIntegral i) 
drawCenterCircle screen_state = arcSolOrange <> arcSol <> triangle <> border <> (getMainChPic screen_state)  <> fstrectangle <> sndrectangle --text = scale 0.3 0.3 $ rotate (-90) $ translate (-400) 0 $ Text $ name $ main_character screen_state 
              where            
                     r = centerCircleRadius screen_state
                     border = color (greyN 0.1) $ thickCircle (r-3) 6
                     arcSol = color (greyN 0.2) $ arcSolid 195 345 r
                     arcSolOrange = (color orange) $ circleSolid r
                     triangle = color orange $ polygon [(0,0) , (r*cos angle,-r* sin angle),(-r* cos angle,-r* sin angle)]
                     (x,y) = polar2decart (2*r/(-3), 3*pi/8)
                     getMainChPic = (translate x y ). (scale (r*0.001) (r*0.001))  . (rotate (-90)) . picture . main_character
                     angle = pi/12
                     fstrectangle = (color$  greyN 0.1) $ translate 0 (-r*(1+sin angle)/2) $ rectangleSolid 25 (r*(1 - sin angle))
                     sndrectangle = (color$  greyN 0.1) $ translate 0 (-r*(sin angle)) $ rectangleSolid (2*(cos angle)*centerCircleRadius screen_state) 15

drawSectors screen_state =  coloredSectors ++ lines -- ++ squares
       where  
              n = segments_quantity screen_state
              lines = let fstpoint i =  polar2decart (l + littleCircleRadius screen_state,realToFrac $ radians *(1/2+ fromIntegral i))
                      in [color white $ line [fstpoint i,mulSV 3 $ fstpoint i] | i<- [0..n-1]]
              squares = concat [[  let (x,y) = polar2decart (l*(1.4 + (j/2)**2),realToFrac $ radians *(1/2+ fromIntegral i))
                                       trans = (color orange) . (translate x y) . rotate(-90) . (scale 0.3 0.3) . Text $ "Info"  -- 
                                       unitOrangeSquare =  ( color (dark orange)$ rectangleSolid 1 1) <> ( color orange $ scale 0.9 0.9 $ rectangleSolid 1 1)
                                       square =  (translate x y) . rotate(-90) . (scale (j*100*scaleX screen_state) (j*200*scaleY screen_state)) $ unitOrangeSquare 
                                   in  square
                                   | i <- [0..n-1] ]
                                   | j<-[3] ]
              coloredSectors  =  zipWith (color) colors  backgroundSectors
              backgroundSectors = zipWith (<>) backgroundSectors_ (sectorContentCircles screen_state)  
              backgroundSectors_ = zipWith (<>) (sectorBackgroundCircles screen_state)  thickArcs
              radians = segmentInRadians screen_state -- 2*(pi::Double) /  fromIntegral n
              l =  distanceToLittleCircle screen_state
              degrees = 360 /fromIntegral n
              thickArcs = map ithThickArc [0..n-1]
              ithThickArc i = thickArc (degrees * i'+border_angle/2) (degrees*(i'+1)-border_angle/2) arcRadius (arcThickness screen_state)
                     where i' = fromIntegral i

sectorBackgroundCircles screen_state = [ let (x,y) = polar2decart (l,realToFrac $ radians *(1/2+ fromIntegral i))
                                         in translate x y $ littleCircle
                                          | i <- [0..n-1] ]
       where
              contentCircle = if focusTarget screen_state == SubjectInCenter
                              then color (dark . dark $ green) $ circleSolid $ 0.9 * (littleCircleRadius screen_state)
                              else Blank
              backgroundCircle = circleSolid $ (littleCircleRadius screen_state) * (degrees - border_angle)/degrees
              littleCircle = backgroundCircle <> contentCircle -- <> characterLittleCircle
              l =  distanceToLittleCircle screen_state
              n = segments_quantity screen_state
              degrees = 360 /fromIntegral n
              radians = segmentInRadians screen_state

sectorContentCircles screen_state = [ let (x,y) = ithCircleCoordinates i
                                          r = littleCircleRadius screen_state
                                          scalar =  0.0015 * r
                                          scalseS = scale scalar scalar
                                          txt = color white $ (  rotate (-90) $ (translate (-0.15*r) (-0.4*r)) . scalseS.  Text . name $ (subjects screen_state) !! i)
                                          subjPic = scalseS . rotate (-90) $ picture $ (subjects screen_state) !! i 
                                          in translate x y $ littleCircle <> subjPic <> txt
                                          | i <- [0..n-1] ]
       where
              contentCircle = if focusTarget screen_state `elem` [SubjectsAround,Balanced] 
                              then color (dark . dark $ green) $ circleSolid $ 0.9 * (littleCircleRadius screen_state)
                              else Blank
              characterLittleCircle = color red $ circleSolid $ (min 0.9 $ (fromIntegral n)/25)*(littleCircleRadius screen_state)
              littleCircle = contentCircle <> characterLittleCircle
              ithCircleCoordinates i = case focusTarget screen_state of
                     SubjectInCenter -> polar2decart (l,realToFrac $ radians *(1/2+ fromIntegral i))
                     Balanced -> polar2decart (l,realToFrac $ radians *(1/2+ fromIntegral i))
                     SubjectsAround -> polar2decart (4*l,realToFrac $ radians *(1/2+ fromIntegral i))

              l =  distanceToLittleCircle screen_state
              n = segments_quantity screen_state
              degrees = 360 /fromIntegral n
              radians = segmentInRadians screen_state


centerCircleRadius :: ScreenState -> Float
centerCircleRadius screen_state@(ScreenState width height _ _ _ focTar) = focusScalar * ( area / (2*pi*(fromIntegral n+1)) ) ** (1/2)  :: Float
          where
              focusScalar = case focTar of 
                     SubjectInCenter -> 1.8
                     SubjectsAround -> 1/2
                     Balanced -> 1                 
              n = segments_quantity screen_state 
              area = fromIntegral $ width * height :: Float

segments_quantity screen_state = length $ subjects screen_state
segmentInRadians screen_state = 2*pi / (fromIntegral $ segments_quantity screen_state)
segmentInDegrees screen_state = 360 / (fromIntegral $ segments_quantity screen_state)

littleCircleRadius :: ScreenState -> Float
littleCircleRadius screen_state = l * sin (realToFrac radians/2)
       where l = distanceToLittleCircle screen_state
             radians = segmentInRadians screen_state

distanceToLittleCircle :: ScreenState -> Float
distanceToLittleCircle screen_state = (centerCircleRadius screen_state) / (1 - sin (realToFrac radians/2 ))
       where
              radians = 2*(pi::Double) /  fromIntegral (segments_quantity screen_state)

arcThickness :: ScreenState -> Float
arcThickness screen_state  =2* ( arcRadius - l * cos (realToFrac radians/2))
       where 
              l = distanceToLittleCircle screen_state
              radians = 2*(pi::Double) /  fromIntegral (segments_quantity screen_state)

polar2decart :: (Float,Float) -> (Float,Float)
polar2decart (r,theta) =  (r * cos theta ,r* sin theta)
decart2polar :: (Float,Float) -> (Float,Float)
decart2polar (x,y) = ((x**2+y**2)**(1/2), atan2 x y)